gtn/cpu/shortest.cpp

```diff
@@ -34,70 +34,91 @@ void shortestDistanceGrad(
     Graph& g,
     float output,
     const Graph& deltas,
+    const std::vector<int>& sortedNodes,
     const std::vector<float>& nodeScores,
-    const std::vector<float>& maxScoresCache,
     const std::vector<size_t>& maxArcIdxCache,
     bool tropical) {
-  std::queue<int> computed;
-  std::vector<size_t> degrees(g.numNodes());
+
   std::vector<float> nodeGrads(g.numNodes(), 0.0);
   std::vector<float> arcGrads(g.numArcs(), 0.0);
-  for (auto n = 0; n < g.numNodes(); ++n) {
-    degrees[n] = g.numOut(n);
-  }
-  float curScore = 0.0;
-  float denom = tropical ? 0.0f : std::exp(output - maxScoresCache.back());
-  for (auto n : g.accept()) {
-    if (g.numOut(n) == 0) {
-      computed.push(n);
-    }
-    if (tropical) {
-      curScore = (n == maxArcIdxCache.back()) ? 1.0f : 0.0f;
-    } else {
-      curScore = std::exp(nodeScores[n] - maxScoresCache.back()) / denom;
-    }
-    nodeGrads[n] += curScore;
-  }
 
-  while (!computed.empty()) {
-    auto n = computed.front();
-    computed.pop();
-    denom = tropical ? 0.0f : std::exp(nodeScores[n] - maxScoresCache[n]);
-    for (const auto a : g.in(n)) {
-      auto un = g.srcNode(a);
+  for (int i = sortedNodes.size() - 1; i >= 0; --i) {
+    auto n = sortedNodes[i];
+    for (const auto a : g.out(n)) {
+      auto dn = g.dstNode(a);
+      float curScore;
       if (tropical) {
-        curScore = (a == maxArcIdxCache[n]) ? nodeGrads[n] : 0.0f;
+        curScore = (a == maxArcIdxCache[dn]) ? nodeGrads[dn] : 0.0f;
+      } else if (nodeScores[dn] == kNegInf) {
+        curScore = 0.0f;
       } else {
-        curScore = nodeGrads[n] *
-            std::exp(nodeScores[un] + g.weight(a) - maxScoresCache[n]) / denom;
+        curScore = nodeGrads[dn] *
+            std::exp(nodeScores[n] + g.weight(a) - nodeScores[dn]);
       }
-      nodeGrads[un] += curScore;
-      arcGrads[a] = curScore * deltas.item();
-      if ((--degrees[un]) == 0) {
-        computed.push(un);
+      nodeGrads[n] += curScore;
+      arcGrads[a] = curScore;
+    }
+    if (g.isAccept(n)) {
+      if (tropical) {
+        nodeGrads[n] += (n == maxArcIdxCache.back()) ? deltas.item() : 0.0f;
+      } else if (nodeScores.back() != kNegInf) {
+        nodeGrads[n] += std::exp(nodeScores[n] - output) * deltas.item();
       }
     }
   }
   g.addGrad(std::move(arcGrads));
 }
 
+std::vector<int> topSort(const Graph& g) {
```

Comment:
is it possible to throw an error if there is a loop in the graph.

Reply:
This top sort is much more efficient for graphs that have a lot of edges but it comes at the cost of not being able to do cycle detection. The idea is you don't have to explore every edge but you have to assume the graph is acyclic to make this assumption.